### ISSUE TITLE
Improve start playback delay

### ISSUE THREAD

**Issue by a3kov** [jdf76/plugin.video.youtube#394]
On weak devices (e.g. TV boxes) delay before playback when you click on a video is too long. Even on a relatively fast Amlogic S912 with fast internet connection it can take up to 4-6 seconds to start.
I played a bit with the plugin on my desktop and its also slow there - it takes about 2 seconds.
I noticed that `VideoInfo._method_get_video_info()` is very slow in particular. It takes many requests to open a single video. I see here 2 ways to improve the situation:
1) Optimize page loading speed - use different algo to fetch the info, use faster loader, etc
2) Prefetch some data when browsing a playlist. This is very tricky for many reasons:
- need to balance between availability of data and amount of wasted data / cpu and RAM usage etc
- need to be careful not to create too many requests to avoid getting captcha/ban etc
- `sts` is kind of session token that probably has its own lifetime and we would need to ensure prefetched data lifetime is less than `sts` lifetime (this is just a guess)

**Comment by corehack** [comment]
The way `VideoInfo.get_player_config()` works with `VideoInfo.get_watch_page()` is very bad. It fetches the whole "watch" page which takes about 1 second even on my desktop. I tried many things and the main problem here is the size of response. Do we really need to fetch the whole page here ?
I see it needs `blankSwfConfig` at the bottom of the page, but I guess it's not needed because its config of the Flash player ?
Another thing at the bottom is `ytInitialPlayerResponse` I have no idea if we really need that, so please chime in...
If we don't need the whole response we could parse just the first chunks.
Another way is to avoid parsing this page, looks like its not needed in every page:
https://www.tyrrrz.me/Blog/Reverse-engineering-YouTube

**Comment by corehack** [comment]
The problem with this addon it has so much legacy code and almost no docstrings. It doesn't even follow PEP8.
What exactly is kodion ? The framework. If it's a framework why is it packaged together with the addon ? So many questions and so few clues..

**Comment by anxdpanic** [comment]
Thanks for looking into it, that blog is a bit out dated but still shows general idea. 

If you look at the summary of the blog, it shows 3 requests made. If you look at The “el” parameter section the last sentence you'll see this _could_ be 4 requests.

Yes that request is required, at this point I don't see a way to lessen this. 
The page is required for the required 'sts' param, the javascript player(required to decipher signatures), the license urls for purchases(required for drm), and handling of some age gates.

**Comment by anxdpanic** [comment]
Kodion is a framework of sorts yes, it's been part of the add-on since before I started maintaining. I am working off Bromix's foundation.

**Comment by corehack** [comment]
We can get sts and the player from the embed page. Vast majority of videos are embeddable.
Most people don't need purchases. Age gates could be a problem but I suspect it's also not a problem for most videos.
Looks like right now the plugin uses the slowest route to the stream because of the features used by minority users or met in minority of videos.

**Comment by anxdpanic** [comment]
You need get_video_info or embedded/watch page to tell it allows embedding, and need the sts for the get_video_info request. So for what you are talking about would means an additional embedded player request for non embedded streams, that seems like an even slower route.

**Comment by anxdpanic** [comment]
That's what I was just mentioning by slower route. We might not see many non-embedded streams individually and most are, but there is a large user base that is likely to notice. (channels that don't allow embedding as a rule, etc)

I'm all for improving this area, don't think I'd be willing to sacrifice one for the other though(I'm not ruling it out). The embedded page also doesn't seem to include adaptive_fmts and url_encoded_fmt_stream_map that are used as fallback like the watch page. YouTube does use SPF, so it might be possible to download only the relevant fragments from the watch page. (something I will look into further)

I'm just about ready to wrap up on 6.1.0, the next version will be more focused on improving things instead of additions.

**Comment by anxdpanic** [comment]
So I attempted fail over from embedded page to watch page and I couldn't find a distinct marker to identify if a video is not embeddable causing some videos to fail. 

What I was able to do was add the ```status``` part to the api request for videos which returns an ```embeddable``` value. Using that cached value so far seems to be reliable.

The tests on Kodi 18 (Windows) 32bit;
```
before:
min: 1.5s max: 2.6s median: 1.7s
```
```
after:
min: 0.9s max: 2.6s median: 1.2s
```
The max values were outliers that both required/used the watch page.

This weekend I expect to bump 6.1.0 to stable and get 6.2.0 alpha started with these changes.
